packages/frontend/src/components/VisualizationConfigs/ChartConfigPanel/CustomVisConfigTabs.tsx

```diff
@@ -73,69 +416,217 @@ const CustomVisConfigTabs: React.FC = memo(() => {
+                {isEditorEmpty ? (
+                    <Text
+                        ml="xl"
+                        pos="absolute"
+                        w="330px"
+                        color="gray.5"
+                        sx={{
+                            pointerEvents: 'none',
+                            zIndex: 100,
+                            fontFamily: 'monospace',
+                        }}
+                    >
+                        {`Write some vega lite JSON or select a template. Check our docs for more info and examples.`}
+                    </Text>
+                ) : null}
 
-            <Tabs.Panel value="data">
                 <Editor
                     loading={<Loader color="gray" size="xs" />}
+                    beforeMount={(monaco) => {
+                        loadMonaco(monaco, schemas.current!);
+                        monacoInstanceRef.current = monaco;
+
+                        // Clean up previous provider if it exists
+                        if (completionProviderRef.current) {
+                            console.debug(
+                                'Clearing Monaco completion provider on beforeMount',
+                                completionProviderRef.current,
+                            );
+                            completionProviderRef.current.dispose();
+                        }
+
+                        if (fields)
+                            completionProviderRef.current =
+                                registerCustomCompletionProvider(
+                                    monaco,
+                                    'json',
+                                    fields,
+                                );
+                    }}
                     defaultLanguage="json"
-                    options={{
-                        ...MONACO_DEFAULT_OPTIONS,
-                        readOnly: true,
+                    options={monacoOptions}
+                    value={editorConfig}
+                    onChange={(config) => {
+                        setEditorConfig(config ?? '');
+                    }}
+                    wrapperProps={{
+                        id: 'tooltip-editor-wrapper',
                     }}
-                    defaultValue={JSON.stringify(series, null, 2)}
                 />
-            </Tabs.Panel>
-        </Tabs>
+            </Group>
+            {/*} </Tabs.Panel>
+
+                <Tabs.Panel value="data">
+                    <Editor
+                        loading={<Loader color="gray" size="xs" />}
+                        defaultLanguage="json"
+                        options={{
+                            ...MONACO_DEFAULT_OPTIONS,
+                            readOnly: true,
+                        }}
+                        defaultValue={JSON.stringify(series, null, 2)}
+                    />
+                </Tabs.Panel>
+            </Tabs>*/}
+        </>
     );
 });
```

Comment:
This file is massive. I suggest extracting components, requests, and utilities into separate files to make it easier for the reader to digest

Reply:
You are right, this is done.